server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -0,0 +1,280 @@
+            InvocationHelper invocationHelper = cr.getComponent(InvocationHelper.class);
+            StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
+            PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
+            Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
+
+            boolean keyMarshalling = MediaType.APPLICATION_OBJECT.equals(scm.getValueStorageMediaType());
+            boolean valueMarshalling = MediaType.APPLICATION_OBJECT.equals(scm.getValueStorageMediaType());
+
+            SerializationContextRegistry ctxRegistry = cm.getGlobalComponentRegistry().getComponent(SerializationContextRegistry.class);
+            ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
+            try (InputStream is = zip.getInputStream(zipEntry)) {
+               while (is.available() > 0) {
+                  CacheBackupEntry entry = readMessageStream(serCtx, CacheBackupEntry.class, is);
+                  Object key = keyMarshalling ? unmarshall(entry.key, userMarshaller) : scm.getKeyWrapper().wrap(entry.key);
+                  Object value = valueMarshalling ? unmarshall(entry.value, userMarshaller) : scm.getKeyWrapper().wrap(entry.value);
+                  Metadata metadata = unmarshall(entry.metadata, persistenceMarshaller);
+                  Metadata internalMetadataImpl = new InternalMetadataImpl(metadata, entry.created, entry.lastUsed);
+
+                  PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
+                        internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
+                  cmd.setInternalMetadata(entry.internalMetadata);
+                  invocationHelper.invoke(cmd, 1);
+               }
+            } catch (IOException e) {
+               throw new CacheException(e);
+            }
+         }, "restore-cache-" + cacheName));
+      }
+      return stages.freeze();
+   }
+
+   private CompletionStage<Void> createCacheBackup(String cacheName) {
+      AdvancedCache<?, ?> cache = cm.getCache(cacheName).getAdvancedCache();
+      Configuration configuration = cm.getCacheConfiguration(cacheName);
+
+      // Create the cache backup dir and parents
+      Path cacheRoot = root.resolve(cacheName);
+      mkdirs(cacheRoot);
+
+      CompletionStage<Void> configStage = blockingManager.runBlocking(() -> {
+         // Write configuration file
+         String xmlFileName = configFile(cacheName);
+         Path xmlPath = cacheRoot.resolve(xmlFileName);
+         try (OutputStream os = Files.newOutputStream(xmlPath)) {
+            parserRegistry.serialize(os, cacheName, configuration);
+         } catch (XMLStreamException | IOException e) {
+            throw new CacheException(String.format("Unable to create backup file '%s'", xmlFileName), e);
+         }
+      }, "write-cache-config");
+
+      if (cache.isEmpty())
```

Comment:
Technically this method can block as well. I would say we could use `Cache#sizeAsync` but that requires getting all elements. TBH, until we have a `Cache#isEmptyAsync` I would recommend just removing this if block. Although it would be quite simple to add this method to the cache if you wanted.

Reply:
I'll remove the if block ... this PR is already pretty large